File: android/app/src/main/java/com/universaldownloader/LinkClassifier.kt
package com.universaldownloader

import java.net.URI

internal object LinkClassifier {
    private val audioFirstHosts = setOf(
        "soundcloud.com",
        "bandcamp.com",
        "music.apple.com",
        "open.spotify.com"
    )

    fun isAudioFirst(url: String): Boolean {
        val normalized = url.trim()
        val host = try {
            URI(normalized).host?.lowercase().orEmpty().removePrefix("www.")
        } catch (_: Exception) {
            ""
        }
        if (host.isBlank()) return false
        return audioFirstHosts.any { host == it || host.endsWith(".$it") }
    }
}

File: ios/Shared/LinkClassifier.swift
import Foundation

enum LinkClassifier {
    private static let audioFirstHosts = [
        "soundcloud.com",
        "bandcamp.com",
        "music.apple.com",
        "open.spotify.com"
    ]

    static func isAudioFirst(_ rawURL: String) -> Bool {
        guard
            let host = URL(string: rawURL.trimmingCharacters(in: .whitespacesAndNewlines))?
                .host?
                .lowercased()
                .replacingOccurrences(of: "www.", with: "")
        else { return false }

        return audioFirstHosts.contains { host == $0 || host.hasSuffix(".\($0)") }
    }
}
